sw/otbn/crypto/sha512.s

```diff
@@ -0,0 +1,967 @@
+      /* w4[63:0] = d <= c */
+      /* w3[63:0] = c <= b */
+      /* w2[63:0] = b <= a */
+      /* w1[63:0] = a = T_1 + T_2 = w20[63:0] + w17[63:0] */
+      bn.add    w1, w20, w17
+
+
+      /* Process word 7 of loop cycle: t <= i*8 + 3. */
+
+      /* w15[255:192] = S0(a) = (a ROTR 28) XOR (a ROTR 34) XOR (a ROTR 39) */
+      bn.rshi  w22,  w1, w30  >> 64
+      bn.rshi  w15,  w1, w22  >> 28
+      bn.rshi  w21,  w1, w22  >> 34
+      bn.xor   w15, w15, w21
+      bn.rshi  w21,  w1, w22  >> 39
+      bn.xor   w15, w15, w21
+
+      /* w16[63:0] = Maj(a,b,c) = (a AND b) XOR (a AND c) XOR (b AND c) */
+      bn.and   w16,  w1,  w2
+      bn.and   w21,  w1,  w3
+      bn.xor   w16, w16, w21
+      bn.and   w21,  w2,  w3
+      bn.xor   w16, w16, w21
+
+      /* w17[63:0] <= T2 = S0(a) + Maj(a,b,c) = w15[255:192] + w16[63:0] */
+      bn.rshi  w17, w30, w15  >> 192
+      bn.add   w17, w17, w16
+
+      /* w18[255:192] <= S1(e) = (e ROTR 14) XOR (e ROTR 18) XOR (e ROTR 41) */
+      bn.rshi  w22,  w5, w30  >> 64
+      bn.rshi  w18,  w5, w22  >> 14
+      bn.rshi  w21,  w5, w22  >> 18
+      bn.xor   w18, w18, w21
+      bn.rshi  w19,  w5, w22  >> 41
+      bn.xor   w18, w18, w19
+
+      /* w19[63:0] <= Ch(e,f,g) = (e AND f) XOR ((NOT e) AND g) */
+      bn.and   w19,  w5,  w6
+      bn.not   w21,  w5
+      bn.and   w21, w21,  w7
+      bn.xor   w19, w19, w21
+
+      /* w20[63:0] <= T1 = h + S1(e) + Ch(e,f,g) + K_t + W_t */
+      bn.rshi  w20, w30, w18  >> 192
+      bn.add   w20, w20,  w0
+      bn.add   w20, w20, w10  >> 192
+      bn.rshi  w21, w30, w11  >> 192
+      bn.add   w21, w21, w19
+      bn.add   w20, w20, w21
+
+      /* w7[63:0] = h <= g */
+      /* w6[63:0] = g <= f */
+      /* w5[63:0] = f <= e */
+      /* w4[63:0] = e <= d + T1 = w2[63:0] + w20[63:0] */
+      bn.add w4, w4, w20
+      /* w3[63:0] = d <= c */
+      /* w2[63:0] = c <= b */
+      /* w1[63:0] = b <= a */
+      /* w0[63:0] = a = T_1 + T_2 = w20[63:0] + w17[63:0] */
+      bn.add w0, w20, w17
+
+
+
+    /* Add compressed chunk to current hash value */
+    addi      x2, x0, 15
+
+    /* H_0 <= H_0 + a */
+    bn.lid    x2, 0(x17)
+    bn.add   w15, w0, w15
+    bn.sid    x2, 0(x17)
+
+    /* H_1 <= H_1 + b */
+    bn.lid    x2, 32(x17)
+    bn.add   w15, w1, w15
+    bn.sid    x2, 32(x17)
+
+    /* H_2 <= H_2 + c */
+    bn.lid    x2, 64(x17)
+    bn.add   w15, w2, w15
+    bn.sid    x2, 64(x17)
+
+    /* H_3 <= H_3 + d */
+    bn.lid    x2, 96(x17)
+    bn.add   w15, w3, w15
+    bn.sid    x2, 96(x17)
+
+    /* H_4 <= H_4 + e */
+    bn.lid    x2, 128(x17)
+    bn.add   w15, w4, w15
+    bn.sid    x2, 128(x17)
+
+    /* H_5 <= H_5 + f */
+    bn.lid    x2, 160(x17)
+    bn.add   w15, w5, w15
+    bn.sid    x2, 160(x17)
+
+    /* H_6 <= H_6 + g */
+    bn.lid    x2, 192(x17)
+    bn.add   w15, w6, w15
+    bn.sid    x2, 192(x17)
+
+    /* H_7 <= H_7 + h */
+    bn.lid    x2, 224(x17)
+    bn.add   w15, w7, w15
+    bn.sid    x2, 224(x17)
+
+ret
+
+.data
```

Comment:
I think all of these except for `K` can go in `.bss` to save a little binary space.